File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
     map<int,int> lrsubtree;// to divide into left and right sub trees
     for(int i=0;i<inorder.size();i++){
        lrsubtree[inorder[i]]=i;
     }
     return helper(preorder,0,preorder.size()-1,inorder,0,inorder.size()-1,lrsubtree);
    }
    TreeNode* helper(vector<int>& preorder,int ps,int pe,vector<int>& inorder,int is, int ie, map<int,int>& lr){
        if( ps > pe || is > ie) return NULL;
        TreeNode* root = new TreeNode(preorder[ps]);
        int rootindex = lr[root->val];// to divide as left subtree and right subtree;
        int lsize = rootindex-is;
        root->left = helper(preorder,ps+1,ps+lsize,inorder,is,rootindex-1,lr);
        root->right = helper(preorder,ps+lsize+1,pe,inorder,rootindex+1,ie,lr);
        return root;
    } 
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
     //pre root left right
     //inorder left root right
     Map<Integer,Integer> mp = new HashMap<>();
     for(int i=0;i<inorder.length;i++)
     {
        mp.put(inorder[i],i);
     }
        return Helper(0,preorder.length-1,preorder,0,inorder.length-1,inorder,mp);
    }
    public TreeNode Helper(int ps, int pe, int[] pre, int is, int ie,int[] in,Map<Integer,Integer> mp){
       if(ps>pe || is>ie) return null;
       TreeNode root = new TreeNode(pre[ps]);
       int rootIndex = mp.get(root.val);
       int leftSize = rootIndex-is;
       root.left = Helper(ps+1,ps+leftSize,pre,is,rootIndex-1,in,mp);
       root.right = Helper(ps+leftSize+1,pe,pre,rootIndex+1,ie,in,mp);
       return root;
       
    }
}